Prevent signed integer overflow in pixel_shuffle size calculation

GregoryComer opened the pull request:
Summary:
The size calculation in `get_pixel_shuffle_out_target_size` can trigger signed integer overflow (UB) with very large upscale_factors. This manifested as an integer division by zero fault during fuzzing.

Specifically, the calculation of `upscale_factor * upscale_factor` ([source](https://github.com/pytorch/executorch/blob/04f1e4d22383ffcbc770acf5002348e3f95082a2/kernels/portable/cpu/util/copy_ops_util.cpp#L364)) can overflow. In the motivating case, SizesType is a signed 32-bit integer and upscale_factor = 2^17.

Since this is an impractically large upscale factor, I'm just adding a constraint that upscale_factor < 32768 (2^15). In theory, SizesType could be defined as less than 32 bits, but this seems unlikely in practice. I check this upper bound in `check_pixel_shuffle_args` and also assert in `get_pixel_shuffle_out_target_size` to ensure we don't hit UB.

Differential Revision: D88693324

JacobSzwejbka commented on kernels/portable/cpu/util/copy_ops_util.cpp:
```diff
@@ -351,6 +353,10 @@ void get_pixel_shuffle_out_target_size(
     size_t* out_ndim) {
+  // Prevent signed integer overflow when computing upscale_factor ^ 2.
+  // This constraint is checked in check_pixel_shuffle_args.
+  ET_CHECK(upscale_factor < 32768);
```
Nit: Can you update this to not fatal error and instead just set the runtime context state

GregoryComer replied:
Yeah, I did add a non-fatal check in the pixel shuffle args check, which will in practice handle it gracefully. I added this assert here just as a sanity check, but am open to changing it if you prefer. It seemed harder to bail out at this point but I wanted to catch any UB.

GregoryComer replied:
Actually it does seems simpler to just refactor the get out size calculation to be able to fail non-fatally and move the check there. I'll push another revision.

GregoryComer committed:
Prevent signed integer overflow in pixel_shuffle size calculation (#16138)

Summary:

The size calculation in `get_pixel_shuffle_out_target_size` can trigger signed integer overflow (UB) with very large upscale_factors. This manifested as an integer division by zero fault during fuzzing.

Specifically, the calculation of `upscale_factor * upscale_factor` ([source](https://github.com/pytorch/executorch/blob/04f1e4d22383ffcbc770acf5002348e3f95082a2/kernels/portable/cpu/util/copy_ops_util.cpp#L364)) can overflow. In the motivating case, SizesType is a signed 32-bit integer and upscale_factor = 2^17.

Since this is an impractically large upscale factor, I'm just adding a constraint that upscale_factor < 32768 (2^15). In theory, SizesType could be defined as less than 32 bits, but this seems unlikely in practice.

Differential Revision: D88693324
